app/src/main/java/com/github/libretube/ui/models/HomeViewModel.kt

```diff
@@ -0,0 +1,170 @@
+                )
+                _loadedSuccessfully.value = trending.value.isNullOrEmpty() == false
+                _isLoading.value = false
+            }
+
+            withContext(Dispatchers.IO) {
+                delay(UNUSUAL_LOAD_TIME_MS)
+                if (result.isActive) {
+                    onUnusualLoadTime.invoke()
+                }
+            }
+        }
+    }
+    private suspend fun loadTrending(context: Context) {
+        val region = LocaleHelper.getTrendingRegion(context)
+
+        runSafely(
+            onSuccess = { videos -> _trending.updateIfChanged(videos) },
+            ioBlock = { RetrofitInstance.api.getTrending(region).deArrow().take(10) }
+        )
+    }
+
+    private suspend fun loadFeed(savedFeed: List<StreamItem>? = null) {
+        runSafely(
+            onSuccess = { videos -> _feed.updateIfChanged(videos) },
+            ioBlock = { tryLoadFeed(savedFeed) }
+        )
+    }
+
+    private suspend fun loadBookmarks() {
+        runSafely(
+            onSuccess = { bookmarks -> _bookmarks.updateIfChanged(bookmarks) },
+            ioBlock = { DatabaseHolder.Database.playlistBookmarkDao().getAll() }
+        )
+    }
+
+    private suspend fun loadPlaylists() {
+        runSafely(
+            onSuccess = { playlists -> _playlists.updateIfChanged(playlists) },
+            ioBlock = { PlaylistsHelper.getPlaylists().take(20) }
+        )
+    }
+
+    private suspend fun loadVideosToContinueWatching() {
+        if (!PlayerHelper.watchHistoryEnabled) return
+        runSafely(
+            onSuccess = { videos -> _continueWatching.updateIfChanged(videos) },
+            ioBlock = ::loadWatchingFromDB
+        )
+    }
+
+    private suspend fun loadWatchingFromDB(): List<StreamItem> {
+        val videos = DatabaseHolder.Database.watchHistoryDao().getAll()
+        return DatabaseHelper
+            .filterUnwatched(videos.map { it.toStreamItem() })
+            .reversed()
+            .take(20)
+    }
+
+    private suspend fun tryLoadFeed(savedFeed: List<StreamItem>?): List<StreamItem> {
+        val feed = if (useSavedFeed && !savedFeed.isNullOrEmpty()) {
+            savedFeed
+        } else {
+            SubscriptionHelper.getFeed()
+        }
+
+        return if (hideWatched) feed.filterWatched() else feed
+    }
+
+    private suspend fun List<StreamItem>.filterWatched(): List<StreamItem> {
+        val allowShorts = ContentFilter.SHORTS.isEnabled()
+        val allowVideos = ContentFilter.VIDEOS.isEnabled()
+        val allowAll = (!allowShorts && !allowVideos)
+
+        val filteredFeed = this.filter {
+            allowAll || (allowShorts && it.isShort) || (allowVideos && !it.isShort)
+        }
+        return runBlocking { DatabaseHelper.filterUnwatched(filteredFeed) }
+    }
+
+    private fun <T> MutableLiveData<T>.updateIfChanged(newValue: T) {
+        if (value != newValue) value = newValue
+    }
```

Comment:
Please move that to that extensions package too, we certainly can re-use this function in other places of the app where view models are being used too :+1: